pygama/dsp/_processors/template.py

```diff
@@ -0,0 +1,77 @@
+# 1) Import python modules
+#
+import numpy as np
+from numba import guvectorize
+
+# 2) Provide instructions to numba
+# 
+# Documentation about numba guvectorize decorator:
+# https://numba.pydata.org/numba-doc/latest/user/vectorize.html#the-guvectorize-decorator
+#
+# Notes:
+# - use `nopython=True` and `cache=True`         (@ian, is this correct?)
+# - use 64-bit variables, e.g. int64 and float64 (@all, this is up for discussion)
+# - use [:] for al output parameters             (@ian, is this correct?)
+#
+@guvectorize("void(float64[:], float64, int64, float64[:], int64[:])",
+             "(n),(),()->()", nopython=True, cache=True)
+
+# 3) Define the processor interface
+#
+# Notes: (@all, this is up for discussion)
+# - add the `_in`/`_out` suffix to the name of the input/output variables
+# - use w_ for waveforms, t_ for indexes, a_ for amplitudes 
+# - use underscore casing for the name of the processor and variables, e.g.
+#   a_trap_energy_in or t_trigger_in
+#
+def the_processor_template(w_in, t_in, a_in, w_out, t_out):
+
+    # 4) Document the algorithm
+    #
+    """
+    Add here a complete description of what the processor does, including the
+    meaning of input and output variables
+    """
+
+    # 5) Initialize output parameters (@iann, will this be done automatically by the processing chain?)
+    #
+    # Notes:
+    # - all output parameters should be initializes to a not-a-number. If a
+    #   processor fails, its output parameters should have the default nan value
+    # - use np.nan for both variables and arrays
+    #
+    w_out[:] = np.nan # use [:] for arrays
+    t_out[0] = np.nan # use [0] for scalar parameters
+
+    # 6) Check inputs. 
+    #
+    # There are two kinds of checks:
+    # - NaN checks. A processor might depend on others, i.e. its input
+    #   parameters are the output parameters of an other processors. When a
+    #   processor fails, all processors depending on its output should not run.
+    #   Thus, skip this processor if a NaN value is detected and return NaN
+    #   output parameters to propagate the failure throughout the processing chain.
+    # - in range checks. Check if indexes are within 0 and len(waveform),
+    #   amplitudes are positive, etc. A failure of this check implies errors in
+    #   the dsp json config file. Abort the analysis immediately.
+    #
+    if (np.isnan(w_in).any() or t_in == np.nan or a_in == np.nan):
+        return
+
+    if (not  0 <= t_in < len(w_in) or not a_in >= 0):
+        pygama_abort();
```

Comment:
This will be done by raising an exception. I will add a try-catch block around the processors to handle these by printing out useful information that can help with debugging, and is not always available to numba functions

Reply:
Is this already available?